server/pages/index.js

```diff
@@ -375,23 +380,64 @@ module.exports = function() {
 	} );
 
 	if ( config.isEnabled( 'manage/themes/details' ) ) {
-		app.get( '/themes/:theme_slug', function( req, res ) {
-			const context = getDefaultContext( req );
+		app.get( '/themes/:theme_slug/:section?', function( req, res ) {
+			function updateRenderCache( themeSlug ) {
+				wpcom.undocumented().themeDetails( themeSlug, ( error, data ) => {
+					if ( error ) {
+						debug( `Error fetching theme ${ themeSlug } details: `, error.message || error );
+						return;
+					}
+					const themeData = themeDetails.get( themeSlug );
+					if ( ! themeData || ( Date( data.date_updated ) > Date( themeData.date_updated ) ) ) {
+						debug( 'caching', themeSlug );
+						themeDetails.set( themeSlug, data );
+						// update the render cache
+						renderThemeSheet( data );
+					}
+				} );
+			}
 
-			if ( config.isEnabled( 'server-side-rendering' ) ) {
-				i18n.initialize();
+			function renderThemeSheet( theme ) {
+				const context = {};
 				const store = createReduxStore();
+				store.dispatch( {
+					type: ThemesActionTypes.RECEIVE_THEME_DETAILS,
+					themeId: theme.id,
+					themeName: theme.name,
+					themeAuthor: theme.author,
+					themeScreenshot: theme.screenshot,
+					themePrice: theme.price ? theme.price.display : undefined,
+					themeDescription: theme.description,
+					themeDescriptionLong: theme.description_long,
+					themeSupportDocumentation: theme.extended ? theme.extended.support_documentation : undefined,
+				} );
 
 				store.dispatch( setSection( 'themes', { hasSidebar: false, isFullScreen: true } ) );
-				context.initialReduxState = pick( store.getState(), 'ui' );
+				context.initialReduxState = pick( store.getState(), 'ui', 'themes' );
 
-				const primary = <ThemeSheetComponent themeSlug={ req.params.theme_slug } />;
+				const primary = <ThemeDetails id={ theme.id }><ThemeSheet /></ThemeDetails>;
 
-				Object.assign( context, render( (
+				const element = (
 					<ReduxProvider store={ store }>
 						<LayoutLoggedOut primary={ primary } />
 					</ReduxProvider>
-				) ) );
+				);
+				const path = url.parse( req.url ).path;
+				const key = JSON.stringify( element ) + path + JSON.stringify( context.initialReduxState );
```

Comment:
Wouldn't we ideally use the the props that `element` receives (possibly including `date_updated`, as [suggested](https://github.com/Automattic/wp-calypso/pull/3279#issuecomment-191976564)) as key? Which would be congruent with the relevant information for `shouldComponentUpdate`, I think.